File: scripts/ocr_law_feedback.js
ObjC.import('Foundation');
ObjC.import('AppKit');
ObjC.import('Quartz');
ObjC.import('PDFKit');
ObjC.import('Vision');
function run(argv) {
  const doc = $.PDFDocument.alloc.initWithURL($.NSURL.fileURLWithPath(argv[0]));
  const rows=[];
  const count=argv[1] ? Math.min(Number(argv[1]),Number(doc.pageCount)) : Number(doc.pageCount);
  for(let i=0;i<count;i++) {
    const page=doc.pageAtIndex(i);
    const thumbnail=page.thumbnailOfSizeForBox($.NSMakeSize(1600,2200),0);
    const ci=$.CIImage.imageWithData(thumbnail.TIFFRepresentation);
    const request=$.VNRecognizeTextRequest.alloc.init;
    request.recognitionLevel=0;
    request.usesLanguageCorrection=true;
    request.recognitionLanguages=$.NSArray.arrayWithObject($('en-GB'));
    const handler=$.VNImageRequestHandler.alloc.initWithCIImageOptions(ci,$.NSDictionary.dictionary);
    const err=Ref();
    if(!handler.performRequestsError($.NSArray.arrayWithObject(request),err)) throw new Error('local_vision_ocr_failed');
    const lines=[];let confidence=0;
    const observations=request.results;
    for(let j=0;j<Number(observations.count);j++) {
      const candidate=observations.objectAtIndex(j).topCandidates(1).objectAtIndex(0);
      lines.push(ObjC.unwrap(candidate.string));confidence+=Number(candidate.confidence);
    }
    const annotations=[];
    for(let j=0;j<Number(page.annotations.count);j++) {
      const text=ObjC.unwrap(page.annotations.objectAtIndex(j).contents);
      if(text) annotations.push(text);
    }
    rows.push({page:i+1,text:lines.join('\n'),annotation_text:annotations.join('\n'),mean_confidence:observations.count ? confidence/Number(observations.count) : 0});
  }
  return JSON.stringify(rows);
}

File: scripts/ocr_law_feedback.swift
import Foundation
import PDFKit
import Vision
import AppKit

// Local Apple Vision OCR only. Stdout is consumed directly into encrypted
// workspace storage by the Python driver, never into shared guide text.
guard CommandLine.arguments.count == 2,
      let document = PDFDocument(url: URL(fileURLWithPath: CommandLine.arguments[1])) else {
    exit(2)
}
for index in 0..<document.pageCount {
    guard let page = document.page(at: index) else { continue }
    let bounds = page.bounds(for: .mediaBox)
    let scale = 2.0
    let image = page.thumbnail(of: NSSize(width: bounds.width * scale, height: bounds.height * scale), for: .mediaBox)
    guard let cgImage = image.cgImage(forProposedRect: nil, context: nil, hints: nil) else { exit(3) }
    let request = VNRecognizeTextRequest()
    request.recognitionLevel = .accurate
    request.usesLanguageCorrection = true
    request.recognitionLanguages = ["en-GB"]
    do {
        try VNImageRequestHandler(cgImage: cgImage, options: [:]).perform([request])
        let observations = request.results ?? []
        let text = observations.compactMap { $0.topCandidates(1).first?.string }.joined(separator: "\n")
        let confidence = observations.compactMap { $0.topCandidates(1).first?.confidence }
        let row: [String: Any] = ["page": index + 1, "text": text,
            "annotation_text": page.annotations.compactMap { $0.contents }.joined(separator: "\n"),
            "mean_confidence": confidence.isEmpty ? 0 : confidence.reduce(0,+) / Float(confidence.count)]
        let data = try JSONSerialization.data(withJSONObject: row, options: [.sortedKeys])
        print(String(data: data, encoding: .utf8)!)
        fflush(stdout)
    } catch { exit(4) }
}
